### DISCUSSION TITLE
How to use SPI_DT_SPEC_GET?

### DISCUSSION THREAD

**Discussion by KenthJohan** [zephyrproject-rtos/zephyr#44490] [Q&A] [answered]
How to use SPI_DT_SPEC_GET?
https://docs.zephyrproject.org/latest/reference/peripherals/spi.html?highlight=spi_dt_spec#c.SPI_DT_SPEC_GET
https://docs.zephyrproject.org/latest/reference/peripherals/spi.html?highlight=spi_dt_spec#c.SPI_DT_SPEC_INST_GET

I cant find examples.
I'm testing on `nucleo_wb55rg`.

I added this overlay so a cs-gpio pin is defined:
**boards/nucleo_wb55rg.overlay**
```overlay
&spi1 {
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5
		     &spi1_miso_pa6 &spi1_mosi_pa7>;
	pinctrl-names = "default";
	status = "okay";
	cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
};
```

I'm not sure why `SPI_DT_SPEC_GET` doesn't work because what I have done is expend it manually with code.
**main.c**
```c
struct spi_dt_spec spi_spec = 
{
	.bus = DEVICE_DT_GET(DT_NODELABEL(spi1)),
	.config = 
	{
		.operation = SPI_WORD_SET(8) | SPI_MODE_GET(0),
		.frequency = 1*1000*1000,
		.cs = &(struct spi_cs_control) 
		{
			.gpio_dev = DEVICE_DT_GET(DT_NODELABEL(gpioa)),
			.delay = 1,
			.gpio_pin = 4,
			.gpio_dt_flags = GPIO_ACTIVE_LOW
		}
	}
};
```
This works well but I would like to replace that with this shortcut so it depends more on the overlay file:
```c
struct spi_dt_spec spi_spec1 = SPI_DT_SPEC_GET(DT_NODELABEL(spi1), SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
```
but that results in:
`identifier "__device_dts_ord_DT_N_S_soc_S_spi_40013000_BUS_ORD" is undefined`
I have this in `devicetree_unfixed.h` but there is no `BUS` word there.
```c
#define DT_N_S_soc_S_spi_40013000_ORD 63
```

**Comment by erwango** [comment, marked as answer]
`SPI_DT_SPEC_GET` should actually be applied to a SPI device (sensor, ...) and not a SPI bus (`spi1` in your case):

Cf:
```
 * @param node_id Devicetree node identifier for the SPI device whose
 *                struct spi_dt_spec to create an initializer for
 * @param operation_ the desired @p operation field in the struct spi_config
 * @param delay_ the desired @p delay field in the struct spi_config's
 *               spi_cs_control, if there is one
 */
#define SPI_DT_SPEC_GET(node_id, operation_, delay_)		     \
```

To use this macro, you should add such device in your dts description.

**Reply by KenthJohan** [reply]
Thank you.
This is the only example I could find right now: https://wiki.st.com/stm32mpu/wiki/SPI_device_tree_configuration
It looks like you can create any device under a SPI bus, but I'm not sure:
```overlay
&spi1 {
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5
		     &spi1_miso_pa6 &spi1_mosi_pa7>;
	pinctrl-names = "default";
	status = "okay";
	cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
	foo@0 {
		compatible = "spi";
		reg = <0>; /* CS #0 */
		spi-max-frequency = <10000000>;
	};
};
```
This generates some extra foo defines in `devicetree_unfixed.h`:
```c
#define DT_N_S_soc_S_spi_40013000_S_foo_0_BUS_spi 1
#define DT_N_S_soc_S_spi_40013000_S_foo_0_BUS DT_N_S_soc_S_spi_40013000
```
But it's still not enough to satisfy `SPI_DT_SPEC_GET`:
```c
struct spi_dt_spec spec = SPI_DT_SPEC_GET(foo, SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
//
```
Its looking for `__device_dts_ord_DT_N_S_soc_S_spi_40013000_BUS_ORD`.
Its actually a quite close match to `DT_N_S_soc_S_spi_40013000` if we remove the `__device_dts_ord_` and `ORD`

What would be the next step in this case? I think I'm missing some macros somewhere.

**Reply by erwango** [reply]
Ok, please note that https://wiki.st.com/stm32mpu/wiki/SPI_device_tree_configuration is specific to STM32MPU configuration for Linux, so Zephyr version will may diverge at some points.
For an example of SPI device configuration, you can have a check to in tree examples, for instance:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/disco_l475_iot1/disco_l475_iot1.dts#L151-L179

Then, you can see how these are used in matching drivers:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/eswifi/eswifi_bus_spi.c#L33
or
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/bluetooth/hci/spi.c#L124

**Reply by KenthJohan** [reply]
I have copied SPI device configuration from https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/disco_l475_iot1/disco_l475_iot1.dts#L151-L179 into `boards/nucleo_wb55rg.overlay` and renamed it to`foo@0` and change configurations.
```c
&spi1 {
	status = "okay";
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5 &spi1_miso_pa6 &spi1_mosi_pa7>;
	pinctrl-names = "default";
	cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
	foo@0 {
		compatible = "zephyr,spi";
		reg = <0>;
		spi-max-frequency = <2000000>;
		label = "FOO";
	};
};
```
This will generate following in `device_treeunfixed.h`
```c
/*
 * Devicetree node: /soc/spi@40013000/foo@0
 *
 * Node identifier: DT_N_S_soc_S_spi_40013000_S_foo_0
 */

/* Node's full path: */
#define DT_N_S_soc_S_spi_40013000_S_foo_0_PATH "/soc/spi@40013000/foo@0"

/* Node's name with unit-address: */
#define DT_N_S_soc_S_spi_40013000_S_foo_0_FULL_NAME "foo@0"

/* Node parent (/soc/spi@40013000) identifier: */
#define DT_N_S_soc_S_spi_40013000_S_foo_0_PARENT DT_N_S_soc_S_spi_40013000
```

In `main.c` file I'm trying to use that configuration similar as https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/wifi/eswifi/eswifi_bus_spi.c#L33

```c
struct spi_dt_spec spec = SPI_DT_SPEC_INST_GET(0, SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
```
The expansion of that macro generates: `__device_dts_ord_DT_N_INST_0_DT_DRV_COMPAT_BUS_ORD`
Which doesn't match anything in `device_treeunfixed.h`.
I have a foo device so I should be able to use it somehow. What am I missing?

**Reply by erwango** [reply]
Sorry, but your question is not clear. You're not supposed to access #defines generated in devicetree_unfixed directly, but through DT macros.
`struct spi_dt_spec spec = SPI_DT_SPEC_GET(DT_NODELABEL(foo), SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);` returns a 'spec' object that you should use with zephyr apis.
Please check documentation on `SPI_DT_SPEC_GET` and `SPI_DT_SPEC_INST_GET` for more information.

**Reply by KenthJohan** [reply]
I know that I can't access it directly that's why I looked in the examples it shows that you can access it with
```c
struct spi_dt_spec spec = SPI_DT_SPEC_INST_GET(0, SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
```
But this doesn't work and I can't figure out why.
The documentation says I should be able to do it but it doesn't work.
Here is the full `main.c` file

```c
#include <drivers/gpio.h>
#include <drivers/spi.h>
#include <drivers/pinctrl.h>
#include <zephyr.h>
#include <device.h>
#include <devicetree.h>

struct spi_dt_spec spec = SPI_DT_SPEC_INST_GET(0, SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);

void main(void)
{
	while (1)
	{
		k_sleep(K_MSEC(1000));
	}
}
```
The compiler error says:
`identifier "__device_dts_ord_DT_N_INST_0_DT_DRV_COMPAT_BUS_ORD" is undefined`
Probably because I'm not using DT macros properly as you said.
I don't know what I'm doing wrong so I can write a clear question.
Either the overlay is wrong which generate incorrect macros or I'm not using DT macros properly.
Sorry I can't pinpoint the exact problem. I hope this makes it more clear.

**Reply by erwango** [reply]
When using `_INST_` DT macros, you should define `DT_DRV_COMPAT`, for instance: 
`#define DT_DRV_COMPAT inventek_eswifi`
Otherwise you should use `SPI_DT_SPEC_GET` version and use DT_NODELABEL(foo) as arugment. Please have check to documentation for more information.

**Reply by KenthJohan** [reply]
Okay now I have this and I still get errors.
**main.c**
```c
#include <drivers/gpio.h>
#include <drivers/spi.h>
#include <drivers/pinctrl.h>
#include <zephyr.h>
#include <device.h>
#include <devicetree.h>

#if 1
struct spi_dt_spec spec = SPI_DT_SPEC_GET(DT_NODELABEL(foo), SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
// error: '__device_dts_ord_DT_N_NODELABEL_foo_BUS_ORD' undeclared
#endif

#if 0
#define DT_DRV_COMPAT foo
struct spi_dt_spec spec = SPI_DT_SPEC_INST_GET(0, SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
// error: '__device_dts_ord_DT_N_INST_0_foo_BUS_ORD' undeclared
#endif


void main(void)
{
	while (1)
	{
		k_sleep(K_MSEC(1000));
	}
}
```
**boards/nucleo_wb55rg.overlay**
```
&spi1 {
	status = "okay";
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5 &spi1_miso_pa6 &spi1_mosi_pa7>;
	pinctrl-names = "default";
	cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
	foo@0 {
		compatible = "zephyr,spi";
		reg = <0>;
		spi-max-frequency = <2000000>;
		label = "FOO";
	};
};
```
It looks so easy in the documentation but in practical its not working.

**Reply by erwango** [reply]
As previously mentioned, I suggest not using DT _INST_ version, unless you know what you're doing.

Here, you're not defining correctly `DT_DRV_COMPAT `, you should define, in your case:
`#define DT_DRV_COMPAT zephyr_spi`

**Reply by KenthJohan** [reply]
Ok now it makes more sense. compatible value is the real identifier.
I stumbled on another error after fixing the first error.
`error: 'DT_N_S_soc_S_spi_40013000_S_foo_0_P_spi_max_frequency' undeclared here`
Which is strange because spi-max-frequency is defined in the overlay file.
```c#
&spi1 {
	status = "okay";
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5 &spi1_miso_pa6 &spi1_mosi_pa7>;
	pinctrl-names = "default";
	cs-gpios = <&gpioa 4 GPIO_ACTIVE_LOW>;
	foo: foo@0 {
		compatible = "zephyr,foo";
		reg = <0>;
		spi-max-frequency = <2000000>;
		label = "foo";
	};
};
```
main.c
```c
#include <drivers/gpio.h>
#include <drivers/spi.h>
#include <drivers/pinctrl.h>
#include <zephyr.h>
#include <device.h>
#include <devicetree.h>

struct spi_dt_spec spec = SPI_DT_SPEC_GET(DT_NODELABEL(foo), SPI_WORD_SET(8) | SPI_MODE_GET(0), 1);
// error: 'DT_N_S_soc_S_spi_40013000_S_foo_0_P_spi_max_frequency' undeclared here

void main(void)
{
	while (1)
	{
		k_sleep(K_MSEC(1000));
	}
}
```

**Reply by erwango** [reply]
Ok, issue is that "zephyr,foo" was not defined as a valid compatible.
So, you need to populate a matching binding, for instance: `dts/bindings/sensor/zephyr,foo.yaml`
with:
```
description: My spi device binding

compatible: "zephyr,foo"

include: spi-device.yaml
```

**Reply by KenthJohan** [reply]
Yeah that works. Thanks.
I'm not sure if this is the correct way to use overlay files.
I want the SPI pins and CS-GPIO pins to be configurable so it can work on multiple boards.
So do we need a custom zephyr repo with a `dts/bindings/sensor/zephyr,foo.yaml` file or is there any better way to do this?

**Reply by erwango** [reply]
If you're wanting to build your own project on top of zephyr repo, adding your own bindings and so on, I'd suggest looking at https://github.com/zephyrproject-rtos/example-application which is provided for that purpose.
